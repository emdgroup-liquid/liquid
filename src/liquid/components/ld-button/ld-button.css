ld-button {
  display: inline-flex;
}

.ld-button:not(.hydrated),
.ld-button.hydrated > button {
  --ld-size-icon-sm: 1rem;
  --ld-size-icon-md: 1.25rem;
  --ld-size-icon-lg: 1.5rem;
  --ld-padding-md: 0.625rem;
  --ld-gap-lg: 0.625rem;
  font: var(--ld-typo-body-m);
  border: 0;
  padding: var(--ld-padding-md) var(--ld-padding-md);
  border-radius: var(--ld-sp-4);
  display: inline-flex;
  min-height: 2.5rem;
  min-width: 2.5rem;
  text-decoration: none;
  user-select: none;
  touch-action: manipulation;
  -webkit-touch-callout: none;
  color: var(--ld-col-wht);
  /*-webkit-tap-highlight-color: transparent;*/

  &:active,
  &:active:focus-visible {
    color: var(--ld-col-wht);
  }

  &:disabled:not(.ld-button--ghost) {
    box-shadow: inset 0px 0px 0px 2px var(--ld-col-disabled);
    color: var(--ld-col-disabled);
  }

  &:not(:disabled) {
    cursor: pointer;
  }

  ld-icon {
    line-height: 0;
  }

  .ld-icon,
  .ld-icon svg {
    width: var(--ld-size-icon-md);
    height: var(--ld-size-icon-md);
  }

  &.ld-button--sm:not(.hydrated),
  &.ld-button--sm.hydrated > button {
    min-height: 2rem;
    min-width: 2rem;
    padding: var(--ld-sp-8) var(--ld-sp-8);
    font: var(--ld-typo-body-s);

    .ld-button__content {
      gap: var(--ld-sp-6);
    }

    .ld-icon,
    .ld-icon svg {
      width: var(--ld-size-icon-sm);
      height: var(--ld-size-icon-sm);
    }
  }

  &.ld-button--lg:not(.hydrated),
  &.ld-button--lg.hydrated > button {
    min-height: 3.125rem;
    min-width: 3.125rem;
    padding: var(--ld-sp-12) var(--ld-sp-12);
    font: var(--ld-typo-body-l);

    .ld-button__content {
      gap: var(--ld-gap-lg);
    }

    .ld-icon,
    .ld-icon svg {
      width: var(--ld-size-icon-lg);
      height: var(--ld-size-icon-lg);
    }
  }

  &.ld-button--align-text-right {
    .ld-button__content {
      text-align: right;
    }
  }

  &.ld-button--align-text-left {
    .ld-button__content {
      text-align: left;
    }
  }

  &.ld-button--justify-start {
    .ld-button__content {
      justify-content: flex-start;
    }
  }

  &.ld-button--justify-end {
    .ld-button__content {
      justify-content: flex-end;
    }
  }

  &.ld-button--justify-between {
    .ld-button__content {
      justify-content: space-between;
    }
  }
}

.ld-button:not(.hydrated),
[class*='ld-theme'] .ld-button:not(.hydrated),
.ld-button.hydrated > button,
[class*='ld-theme'] .ld-button.hydrated > button {
  &.ld-button--highlight {
    background-color: var(--ld-col-vy-default);
    color: var(--ld-col-rblck-default);

    &:focus:focus-visible {
      background-color: var(--ld-col-vy-focus);
    }

    @media (hover: hover) {
      &:hover {
        background-color: var(--ld-col-vy-hover);
      }
    }

    &:active,
    &:active:focus-visible {
      background-color: var(--ld-col-vy-active);
      color: var(--ld-col-wht);
    }
  }

  &.ld-button--danger {
    background-color: var(--ld-col-rr-default);
    color: var(--ld-col-wht);

    &:focus:focus-visible {
      background-color: var(--ld-col-rr-focus);
    }

    @media (hover: hover) {
      &:hover {
        background-color: var(--ld-col-rr-hover);
        color: var(--ld-col-wht);
      }
    }

    &:active,
    &:active:focus-visible {
      background-color: var(--ld-col-rr-active);
      color: var(--ld-col-wht);
    }
  }

  &.ld-button--secondary {
    :not(:disabled) & {
      background-color: var(--ld-col-wht);
    }
  }

  &.ld-button--ghost {
    :not(:disabled) & {
      background-color: transparent;
    }

    &:active,
    &:active:focus-visible {
      background-color: var(--ld-col-bg-g);
    }

    @media (hover: hover) {
      &:hover {
        background-color: var(--ld-col-bg-g);
      }
    }
  }

  &.ld-button--highlight,
  &.ld-button--secondary,
  &.ld-button--ghost,
  &.ld-button--danger {
    &:disabled,
    &:disabled:hover,
    &:disabled:focus,
    &:disabled:active {
      background-color: transparent;
      color: var(--ld-col-disabled);
    }

    @media (hover: hover) {
      &:disabled:hover {
        background-color: transparent;
        color: var(--ld-col-disabled);
      }
    }
  }

  ld-button & {
    width: 100%;
  }
}

.ld-button.ld-button--secondary {
  &:not(:disabled) {
    box-shadow: inset 0px 0px 0px 2px var(--ld-col-rblck1);

    @media (hover: hover) {
      &:hover {
        box-shadow: inset 0px 0px 0px 2px var(--ld-col-rb5);
      }
    }
    &:active,
    &:active:focus-visible {
      box-shadow: inset 0px 0px 0px 2px var(--ld-col-rb55);
    }
    &:focus:focus-visible {
      box-shadow: inset 0px 0px 0px 2px var(--ld-col-rb3);
    }
  }

  .ld-theme-ocean &,
  [class*='ld-theme'] .ld-theme-ocean & {
    &:not(:disabled) {
      @media (hover: hover) {
        &:hover {
          box-shadow: inset 0px 0px 0px 2px var(--ld-col-rb5);
        }
      }
      &:active,
      &:active:focus-visible {
        box-shadow: inset 0px 0px 0px 2px var(--ld-col-rb55);
      }
      &:focus:focus-visible {
        box-shadow: inset 0px 0px 0px 2px var(--ld-col-rb3);
      }
    }
  }

  .ld-theme-solvent &,
  .ld-theme-bubblegum &,
  .ld-theme-shake &,
  [class*='ld-theme'] .ld-theme-solvent &,
  [class*='ld-theme'] .ld-theme-bubblegum &,
  [class*='ld-theme'] .ld-theme-shake & {
    &:not(:disabled) {
      @media (hover: hover) {
        &:hover {
          box-shadow: inset 0px 0px 0px 2px var(--ld-col-rp5);
        }
      }
      &:active,
      &:active:focus-visible {
        box-shadow: inset 0px 0px 0px 2px var(--ld-col-rp55);
      }
      &:focus:focus-visible {
        box-shadow: inset 0px 0px 0px 2px var(--ld-col-rp3);
      }
    }
  }

  .ld-theme-tea &,
  [class*='ld-theme'] .ld-theme-tea & {
    &:not(:disabled) {
      @media (hover: hover) {
        &:hover {
          box-shadow: inset 0px 0px 0px 2px var(--ld-col-rg5);
        }
      }
      &:active,
      &:active:focus-visible {
        box-shadow: inset 0px 0px 0px 2px var(--ld-col-rg6);
      }
      &:focus:focus-visible {
        box-shadow: inset 0px 0px 0px 2px var(--ld-col-rg3);
      }
    }
  }
}

.ld-button.ld-button--secondary,
.ld-button.ld-button--ghost {
  &:not(:disabled) {
    color: var(--ld-col-rblck-default);

    @media (hover: hover) {
      &:hover {
        color: var(--ld-col-rb5);
      }
    }
    &:active,
    &:active:focus-visible {
      color: var(--ld-col-rb55);
    }
    &:focus:focus-visible {
      color: var(--ld-col-rb3);
    }
  }

  .ld-theme-ocean &,
  [class*='ld-theme'] .ld-theme-ocean & {
    @media (hover: hover) {
      &:hover {
        color: var(--ld-col-rb5);
      }
    }
    &:active,
    &:active:focus-visible {
      color: var(--ld-col-rb55);
    }
    &:focus:focus-visible {
      color: var(--ld-col-rb3);
    }
  }

  .ld-theme-solvent &,
  .ld-theme-bubblegum &,
  .ld-theme-shake &,
  [class*='ld-theme'] .ld-theme-solvent &,
  [class*='ld-theme'] .ld-theme-shake &,
  [class*='ld-theme'] .ld-theme-bubblegum & {
    color: var(--ld-col-rp6);

    @media (hover: hover) {
      &:hover {
        color: var(--ld-col-rp5);
      }
    }
    &:active,
    &:active:focus-visible {
      color: var(--ld-col-rp55);
    }
    &:focus:focus-visible {
      color: var(--ld-col-rp3);
    }
  }

  .ld-theme-tea &,
  [class*='ld-theme'] .ld-theme-tea & {
    @media (hover: hover) {
      &:hover {
        color: var(--ld-col-rg5);
      }
    }
    &:active,
    &:active:focus-visible {
      color: var(--ld-col-rg6);
    }
    &:focus:focus-visible {
      color: var(--ld-col-rg3);
    }
  }
}

.ld-button__content {
  display: grid;
  grid-auto-flow: column;
  gap: var(--ld-sp-8);
  align-items: center;
  width: 100%;
  align-self: center;
  text-align: center;
  justify-content: center;
  font-weight: 900;
  line-height: 1;
}
