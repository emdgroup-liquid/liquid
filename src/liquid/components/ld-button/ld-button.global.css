/* .ld-theme-ocean -> default */
.ld-theme-ocean,
[class*='ld-theme'] .ld-theme-ocean {
  ld-button,
  .ld-button {
    --ld-button-bg-color-active: var(--ld-col-rb-active);
    --ld-button-bg-color-focus: var(--ld-col-rb-focus);
    --ld-button-bg-color-hover: var(--ld-col-rb-hover);
    --ld-button-bg-color: var(--ld-thm-ocean-bg-primary);
    --ld-button-ghost-bg-color-active: var(--ld-col-rb-a02);
    --ld-button-ghost-bg-color-hover: var(--ld-col-rb-a01);
    --ld-button-ghost-progress-bg-color: var(--ld-col-rb-a01);
    --ld-button-ghost-progress-color: var(--ld-col-rb-default);
    --ld-button-ghost-text-color-active: var(--ld-col-rb-active);
    --ld-button-ghost-text-color-focus: var(--ld-col-rb-focus);
    --ld-button-ghost-text-color-hover: var(--ld-col-rb-hover);
    --ld-button-ghost-text-color: var(--ld-col-rb-default);
    --ld-button-progress-color: var(--ld-thm-ocean-accent);
    --ld-button-secondary-shadow-color-active: var(--ld-col-rb-active);
    --ld-button-secondary-shadow-color-focus: var(--ld-col-rb-focus);
    --ld-button-secondary-shadow-color-hover: var(--ld-col-rb-hover);
    --ld-button-secondary-shadow-color: var(--ld-col-rb-default);
    --ld-button-text-brand-color-focus: var(--ld-col-rb-focus);
    --ld-button-text-brand-color: var(--ld-col-rb-default);
    --ld-button-text-color-focus: var(--ld-thm-ocean-bg-secondary);
  }
}

.ld-theme-bubblegum,
[class*='ld-theme'] .ld-theme-bubblegum,
.ld-theme-shake,
[class*='ld-theme'] .ld-theme-shake,
.ld-theme-solvent,
[class*='ld-theme'] .ld-theme-solvent {
  :where(ld-button, .ld-button) {
    --ld-button-bg-color-active: var(--ld-col-rp-active);
    --ld-button-bg-color-focus: var(--ld-col-rp-focus);
    --ld-button-bg-color-hover: var(--ld-col-rp-hover);
    --ld-button-bg-color: var(--ld-thm-solvent-bg-primary);
    --ld-button-ghost-bg-color-active: var(--ld-col-rp-a02);
    --ld-button-ghost-bg-color-hover: var(--ld-col-rp-a01);
    --ld-button-ghost-progress-bg-color: var(--ld-col-rp-a01);
    --ld-button-ghost-progress-color: var(--ld-col-rp-default);
    --ld-button-ghost-text-color-active: var(--ld-col-rp-active);
    --ld-button-ghost-text-color-focus: var(--ld-col-rp-focus);
    --ld-button-ghost-text-color-hover: var(--ld-col-rp-hover);
    --ld-button-ghost-text-color: var(--ld-col-rp-default);
    --ld-button-secondary-shadow-color-active: var(--ld-col-rp-active);
    --ld-button-secondary-shadow-color-focus: var(--ld-col-rp-focus);
    --ld-button-secondary-shadow-color-hover: var(--ld-col-rp-hover);
    --ld-button-secondary-shadow-color: var(--ld-col-rp-default);
    --ld-button-text-brand-color-focus: var(--ld-col-rp-focus);
    --ld-button-text-brand-color: var(--ld-col-rp-default);
    --ld-button-text-color-focus: var(--ld-thm-solvent-bg-secondary);
  }
}

.ld-theme-bubblegum,
[class*='ld-theme'] .ld-theme-bubblegum {
  :where(ld-button, .ld-button) {
    --ld-button-progress-color: var(--ld-thm-bubblegum-accent);
  }
}

.ld-theme-shake,
[class*='ld-theme'] .ld-theme-shake {
  :where(ld-button, .ld-button) {
    --ld-button-progress-color: var(--ld-thm-shake-accent);
  }
}

.ld-theme-solvent,
[class*='ld-theme'] .ld-theme-solvent {
  :where(ld-button, .ld-button) {
    --ld-button-progress-color: var(--ld-thm-solvent-accent);
  }
}

.ld-theme-tea,
[class*='ld-theme'] .ld-theme-tea {
  :where(ld-button, .ld-button) {
    --ld-button-bg-color-active: var(--ld-col-rg-active);
    --ld-button-bg-color-focus: var(--ld-col-rg-focus);
    --ld-button-bg-color-hover: var(--ld-col-rg-hover);
    --ld-button-bg-color: var(--ld-thm-tea-bg-primary);
    --ld-button-ghost-bg-color-active: var(--ld-col-rg-a02);
    --ld-button-ghost-bg-color-hover: var(--ld-col-rg-a01);
    --ld-button-ghost-progress-bg-color: var(--ld-col-rg-a01);
    --ld-button-ghost-progress-color: var(--ld-col-rg-default);
    --ld-button-ghost-text-color-active: var(--ld-col-rg-active);
    --ld-button-ghost-text-color-focus: var(--ld-col-rg-focus);
    --ld-button-ghost-text-color-hover: var(--ld-col-rg-hover);
    --ld-button-ghost-text-color: var(--ld-col-rg-default);
    --ld-button-progress-color: var(--ld-thm-tea-accent);
    --ld-button-secondary-shadow-color-active: var(--ld-col-rg-active);
    --ld-button-secondary-shadow-color-focus: var(--ld-col-rg-focus);
    --ld-button-secondary-shadow-color-hover: var(--ld-col-rg-hover);
    --ld-button-secondary-shadow-color: var(--ld-col-rg-default);
    --ld-button-text-brand-color-focus: var(--ld-col-rg-focus);
    --ld-button-text-brand-color: var(--ld-col-rg-default);
    --ld-button-text-color-focus: var(--ld-thm-tea-bg-secondary);
  }
}
