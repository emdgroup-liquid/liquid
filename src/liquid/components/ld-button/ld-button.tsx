import '../../components' // type definitions for type checking and intelliSense
import { Component, h, Host, Prop, Element } from '@stencil/core'
import { cloneAttributes } from '../../utils/cloneAttributes'
import { JSXBase } from '@stencil/core/internal'
import ButtonHTMLAttributes = JSXBase.ButtonHTMLAttributes
import AnchorHTMLAttributes = JSXBase.AnchorHTMLAttributes
import { applyPropAliases } from '../../utils/applyPropAliases'

@Component({
  tag: 'ld-button',
  styleUrl: 'ld-button.css',
  shadow: false,
})
export class LdButton {
  @Element() el: HTMLElement

  private button!: HTMLElement

  /** Disabled state of the button. */
  @Prop() disabled = false

  /** Size of the button. */
  @Prop() size?: 'sm' | 'lg'

  /** Display mode. */
  @Prop() mode?:
    | 'highlight'
    | 'secondary'
    | 'ghost'
    | 'danger'
    | 'on-brand-color'
    | 'secondary-on-brand-color'

  /** Align text. */
  @Prop({ mutable: true }) alignText: 'left' | 'right'

  /** Justify content. */
  @Prop({ mutable: true }) justifyContent: 'start' | 'end' | 'between'

  /**
   * Transforms the button to an anchor element.
   * See [mdn docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-href)
   * for more information on the `href` attribute.
   */
  @Prop() href?: string

  /**
   * The `target` attributed can be used in conjunction with the `href` attribute.
   * See [mdn docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-target)
   * for more information on the `target` attribute.
   */
  @Prop() target?: '_blank' | '_self' | '_parent' | '_top'

  private handleClick(ev) {
    if (this.button.getAttribute('aria-disabled')) {
      ev.preventDefault()
      ev.stopImmediatePropagation()
    }
  }

  componentWillLoad() {
    applyPropAliases.apply(this)
  }

  render() {
    let cl = 'ld-button'
    if (this.size) cl += ` ld-button--${this.size}`
    if (this.mode) cl += ` ld-button--${this.mode}`
    if (this.alignText) cl += ` ld-button--align-text-${this.alignText}`
    if (this.justifyContent) cl += ` ld-button--justify-${this.justifyContent}`

    const Tag = this.href ? 'a' : 'button'

    return (
      <Host>
        <Tag
          ref={(el) => (this.button = el as HTMLElement)}
          onClick={this.handleClick.bind(this)}
          class={cl}
          disabled={this.disabled}
          aria-disabled={this.disabled ? 'true' : undefined}
          href={this.href}
          target={this.target}
          rel={this.target === '_blank' ? 'noreferrer noopener' : undefined}
          {...(this.href
            ? cloneAttributes<ButtonHTMLAttributes<HTMLButtonElement>>(this.el)
            : cloneAttributes<AnchorHTMLAttributes<HTMLAnchorElement>>(
                this.el
              ))}
        >
          <slot />
        </Tag>
      </Host>
    )
  }
}
