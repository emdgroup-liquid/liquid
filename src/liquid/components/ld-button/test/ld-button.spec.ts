import { newSpecPage } from '@stencil/core/testing'
import { LdButton } from '../ld-button'
import '../../../utils/mutationObserver'

const mockClickHiddenButton = (
  form: HTMLFormElement,
  buttonType: 'submit' | 'reset'
) => {
  // Mock clickHiddenButton (actual implementation tested in e2e test).
  jest
    .spyOn(
      LdButton.prototype as unknown as { clickHiddenButton: () => void },
      'clickHiddenButton'
    )
    .mockImplementation(() => {
      form.dispatchEvent(new Event(buttonType))
    })
}

describe('ld-button', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button>Text</ld-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('disabled', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button disabled>Text</ld-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('highlight', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button mode="highlight">Text</ld-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('secondary', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button mode="secondary">Text</ld-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('brand-color (default)', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button brand-color>Text</ld-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('brand-color (secondary)', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button mode="secondary" brand-color>Text</ld-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('brand-color (ghost)', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button mode="ghost" brand-color>Text</ld-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('ghost', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button mode="ghost">Text</ld-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('danger', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button mode="danger">Text</ld-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('size', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button size="sm">Text</ld-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('justify content', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button justify-content="end">Text</ld-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('align text', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button align-text="right">Text</ld-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('as anchor', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button href="#" target="_blank">Text</ld-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('as progress button', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button progress="0.75">Text</ld-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('as progress button pending', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button progress="pending">Text</ld-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('allows adding click handlers', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button disabled>Text</ld-button>`,
    })
    const ldButton = page.root
    const clickHandler = jest.fn()

    ldButton.addEventListener('click', clickHandler)
    ldButton.dispatchEvent(
      new MouseEvent('click', { bubbles: true, cancelable: true })
    )

    expect(clickHandler).toHaveBeenCalledTimes(1)
  })

  xit('prevents click handlers from being invoked, if disabled', async () => {
    // FIXME: the event handlers being called might be a stencil "jsdom" issue
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button disabled>Text</ld-button>`,
    })
    const ldButton = page.root
    const clickHandler = jest.fn()

    ldButton.addEventListener('click', clickHandler)
    ldButton.dispatchEvent(
      new MouseEvent('click', { bubbles: true, cancelable: true })
    )

    expect(clickHandler).not.toHaveBeenCalled()
  })

  xit('prevents click handlers from being invoked, if aria-disabled', async () => {
    // FIXME: the event handlers being called might be a stencil "jsdom" issue
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button aria-disabled="true">Text</ld-button>`,
    })
    const ldButton = page.root
    const clickHandler = jest.fn()

    ldButton.addEventListener('click', clickHandler)
    ldButton.dispatchEvent(
      new MouseEvent('click', { bubbles: true, cancelable: true })
    )

    expect(clickHandler).not.toHaveBeenCalled()
  })

  it('allows to set inner focus', async () => {
    const { root } = await newSpecPage({
      components: [LdButton],
      html: `<ld-button>Text</ld-button>`,
    })
    const button = root.shadowRoot.querySelector('button')

    button.focus = jest.fn()
    await root.focusInner()

    expect(button.focus).toHaveBeenCalled()
  })

  describe('implicit form submission', () => {
    afterAll(() => {
      jest.restoreAllMocks()
    })

    it('submits a form implicitly', async () => {
      const page = await newSpecPage({
        components: [LdButton],
        html: `<form><ld-button>Text</ld-button></form>`,
      })
      const form = page.body.querySelector('form')

      mockClickHiddenButton(form, page.rootInstance.type)

      const ldButton = page.body.querySelector('ld-button')
      const submitHandler = jest.fn()

      form.addEventListener('submit', submitHandler)
      ldButton.dispatchEvent(
        new MouseEvent('click', { bubbles: true, cancelable: true })
      )

      jest.advanceTimersByTime(0)

      expect(submitHandler).toHaveBeenCalled()
    })

    it('does not submit a form as an anchor', async () => {
      const page = await newSpecPage({
        components: [LdButton],
        html: `<form><ld-button href="#">Text</ld-button></form>`,
      })
      const form = page.body.querySelector('form')

      mockClickHiddenButton(form, page.rootInstance.type)

      const ldButton = page.body.querySelector('ld-button')
      const submitHandler = jest.fn()

      form.addEventListener('submit', submitHandler)
      ldButton.dispatchEvent(
        new MouseEvent('click', { bubbles: true, cancelable: true })
      )

      jest.advanceTimersByTime(0)

      expect(submitHandler).not.toHaveBeenCalled()
    })

    it('does not submit a form if event is prevented', async () => {
      const page = await newSpecPage({
        components: [LdButton],
        html: `<form><ld-button>Text</ld-button></form>`,
      })
      const form = page.body.querySelector('form')

      mockClickHiddenButton(form, page.rootInstance.type)

      const ldButton = page.body.querySelector('ld-button')
      const submitHandler = jest.fn()

      form.addEventListener('submit', submitHandler)
      const ev = new MouseEvent('click', { bubbles: true, cancelable: true })
      ldButton.dispatchEvent(ev)
      // preventing after dispatching only works with setTimeout implementation
      // which is required in order to check for ev.defaultPrevented
      ev.preventDefault()

      jest.advanceTimersByTime(0)

      expect(submitHandler).not.toHaveBeenCalled()
    })

    it('resets a form', async () => {
      const page = await newSpecPage({
        components: [LdButton],
        html: `<form><ld-button type="reset">Text</ld-button></form>`,
      })
      const form = page.body.querySelector('form')

      mockClickHiddenButton(form, page.rootInstance.type)

      const ldButton = page.body.querySelector('ld-button')
      const resetHandler = jest.fn()

      form.addEventListener('reset', resetHandler)
      ldButton.dispatchEvent(
        new MouseEvent('click', { bubbles: true, cancelable: true })
      )

      jest.advanceTimersByTime(0)

      expect(resetHandler).toHaveBeenCalled()
    })
  })

  it('removes size from ld-icon web component', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button>
        <ld-icon name="placeholder" size="sm"></ld-icon>
        <ld-icon name="placeholder" size="lg"></ld-icon>
      </ld-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('removes size from ld-icon css component', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button>
        <svg class="ld-icon ld-icon--sm"></svg>
        <svg class="ld-icon ld-icon--lg"></svg>
      </ld-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('sets size on ld-icon web component', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button size="sm">
        <ld-icon name="placeholder"></ld-icon>
        <ld-icon name="placeholder" size="lg"></ld-icon>
      </ld-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('sets size on ld-icon css component', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button size="sm">
        <svg class="ld-icon"></svg>
        <svg class="ld-icon ld-icon--lg"></svg>
      </ld-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('clones attributes to inner button', async () => {
    const page = await newSpecPage({
      components: [LdButton],
      html: `<ld-button size="sm" aria-label="yolo" hidden></ld-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })
})
