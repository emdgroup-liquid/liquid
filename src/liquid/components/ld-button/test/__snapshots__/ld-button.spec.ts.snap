// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-button align text 1`] = `
<ld-button align-text="right">
  <mock:shadow-root>
    <button aria-live="polite" class="ld-button ld-button--align-text-right" part="button focusable">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Text
</ld-button>
`;

exports[`ld-button as anchor 1`] = `
<ld-button href="#" target="_blank">
  <mock:shadow-root>
    <a aria-live="polite" class="ld-button" href="#" part="button focusable" rel="noreferrer noopener" target="_blank">
      <slot></slot>
    </a>
  </mock:shadow-root>
  Text
</ld-button>
`;

exports[`ld-button as progress button 1`] = `
<ld-button progress="0.75">
  <mock:shadow-root>
    <button aria-busy="true" aria-live="polite" class="ld-button" part="button focusable">
      <slot></slot>
      <span class="ld-button__progress" part="progress-bar" style="--ld-button-progress: 0.75;"></span>
    </button>
  </mock:shadow-root>
  Text
</ld-button>
`;

exports[`ld-button as progress button pending 1`] = `
<ld-button progress="pending">
  <mock:shadow-root>
    <button aria-busy="true" aria-live="polite" class="ld-button" part="button focusable">
      <slot></slot>
      <span class="ld-button__progress ld-button__progress--pending" part="progress-bar"></span>
    </button>
  </mock:shadow-root>
  Text
</ld-button>
`;

exports[`ld-button brand-color (default) 1`] = `
<ld-button brand-color="">
  <mock:shadow-root>
    <button aria-live="polite" class="ld-button ld-button--brand-color" part="button focusable">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Text
</ld-button>
`;

exports[`ld-button brand-color (ghost) 1`] = `
<ld-button brand-color="" mode="ghost">
  <mock:shadow-root>
    <button aria-live="polite" class="ld-button ld-button--brand-color ld-button--ghost" part="button focusable">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Text
</ld-button>
`;

exports[`ld-button brand-color (secondary) 1`] = `
<ld-button brand-color="" mode="secondary">
  <mock:shadow-root>
    <button aria-live="polite" class="ld-button ld-button--brand-color ld-button--secondary" part="button focusable">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Text
</ld-button>
`;

exports[`ld-button clones attributes to inner button 1`] = `
<ld-button aria-label="yolo" hidden="" size="sm">
  <mock:shadow-root>
    <button aria-label="yolo" aria-live="polite" class="ld-button ld-button--sm" hidden="" part="button focusable">
      <slot></slot>
    </button>
  </mock:shadow-root>
</ld-button>
`;

exports[`ld-button danger 1`] = `
<ld-button mode="danger">
  <mock:shadow-root>
    <button aria-live="polite" class="ld-button ld-button--danger" part="button focusable">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Text
</ld-button>
`;

exports[`ld-button disabled 1`] = `
<ld-button disabled="">
  <mock:shadow-root>
    <button aria-disabled="true" aria-live="polite" class="ld-button" disabled="" part="button focusable">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Text
</ld-button>
`;

exports[`ld-button ghost 1`] = `
<ld-button mode="ghost">
  <mock:shadow-root>
    <button aria-live="polite" class="ld-button ld-button--ghost" part="button focusable">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Text
</ld-button>
`;

exports[`ld-button highlight 1`] = `
<ld-button mode="highlight">
  <mock:shadow-root>
    <button aria-live="polite" class="ld-button ld-button--highlight" part="button focusable">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Text
</ld-button>
`;

exports[`ld-button justify content 1`] = `
<ld-button justify-content="end">
  <mock:shadow-root>
    <button aria-live="polite" class="ld-button ld-button--justify-end" part="button focusable">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Text
</ld-button>
`;

exports[`ld-button removes size from ld-icon css component 1`] = `
<ld-button>
  <mock:shadow-root>
    <button aria-live="polite" class="ld-button" part="button focusable">
      <slot></slot>
    </button>
  </mock:shadow-root>
  <svg class="ld-icon"></svg>
  <svg class="ld-icon"></svg>
</ld-button>
`;

exports[`ld-button removes size from ld-icon web component 1`] = `
<ld-button>
  <mock:shadow-root>
    <button aria-live="polite" class="ld-button" part="button focusable">
      <slot></slot>
    </button>
  </mock:shadow-root>
  <ld-icon name="placeholder"></ld-icon>
  <ld-icon name="placeholder"></ld-icon>
</ld-button>
`;

exports[`ld-button renders 1`] = `
<ld-button>
  <mock:shadow-root>
    <button aria-live="polite" class="ld-button" part="button focusable">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Text
</ld-button>
`;

exports[`ld-button secondary 1`] = `
<ld-button mode="secondary">
  <mock:shadow-root>
    <button aria-live="polite" class="ld-button ld-button--secondary" part="button focusable">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Text
</ld-button>
`;

exports[`ld-button sets size on ld-icon css component 1`] = `
<ld-button size="sm">
  <mock:shadow-root>
    <button aria-live="polite" class="ld-button ld-button--sm" part="button focusable">
      <slot></slot>
    </button>
  </mock:shadow-root>
  <svg class="ld-icon ld-icon--sm"></svg>
  <svg class="ld-icon ld-icon--sm"></svg>
</ld-button>
`;

exports[`ld-button sets size on ld-icon web component 1`] = `
<ld-button size="sm">
  <mock:shadow-root>
    <button aria-live="polite" class="ld-button ld-button--sm" part="button focusable">
      <slot></slot>
    </button>
  </mock:shadow-root>
  <ld-icon name="placeholder" size="sm"></ld-icon>
  <ld-icon name="placeholder" size="sm"></ld-icon>
</ld-button>
`;

exports[`ld-button size 1`] = `
<ld-button size="sm">
  <mock:shadow-root>
    <button aria-live="polite" class="ld-button ld-button--sm" part="button focusable">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Text
</ld-button>
`;
